### PR TITLE
improve default playback quality of YouTube videos to 1440p (or `vq=`, `quality=` query-string parameter) (fixes issue #1051)

### PR DESCRIPTION
**Pull request by cvan**
for issue #1051

### CODE REVIEW

**Comment by MortimerGoro** [comment]
I tried with this video but it still selects 380 or 480p by default: https://www.youtube.com/watch?v=sPyAQQklc1s

Can you check it?

**Comment by cvan** [comment]
@MortimerGoro I fiddled with it a bit. fixing now, should have it in a good state by tomorrow. will ping you when it's ready to test again. thanks!

**Comment by cvan** [comment]
> This is great, the hard part will be to maintain it when YouTube breaks it.

quite possibly, yes. still tinkering here and there. I think I can cut the code down considerably.

another thought: possibly we can remotely host the CSS + JS so if (I mean when) things break, we don't have to roll out new releases. thoughts?

**Comment by bluemarvin** [comment]
Since we don’t sign the extensions remote hosting would be an exploit vector.

**Comment by cvan** [comment]
> Since we don’t sign the extensions remote hosting would be an exploit vector.

yeah, I agree. CSP isn't enforced on these content scripts (yet; GV bug?), so we're already pushing it heh.

File: app/src/common/shared/org/mozilla/vrbrowser/browser/SessionStore.java [hunk 1]
```diff
@@ -1042,7 +1042,9 @@ public void onCanGoForward(@NonNull GeckoSession aSession, boolean aCanGoForward
     private String checkYoutubeOverride(String aUri) {
         try {
             Uri uri = Uri.parse(aUri);
-            if (!uri.getHost().toLowerCase().contains("www.youtube.")) {
+            String hostLower = !uri.getHost().toLowerCase();
```
Comment (bluemarvin) [inline app/src/common/shared/org/mozilla/vrbrowser/browser/SessionStore.java]: Typo here ~!~

Comment (cvan) [reply app/src/common/shared/org/mozilla/vrbrowser/browser/SessionStore.java]: thanks (been running this WebExt on desktop for rapider development)

**Comment by bluemarvin** [comment]
Web extensions are only intended for internal use right now in GV so there aren’t security measures by design I think.

**Comment by philip-lamb** [comment]
Can we also comment/test whether this works against issues #832 #837 #793.

**Comment by cvan** [comment]
> Can we also comment/test whether this works against issues 
Thanks for checking!

> #832
Nope, unrelated - we can probably close that. I haven't found it necessary.

> #837
Yep, will be handled. I will reference the issue in the commit message when I push it to this PR branch.

> #793 
No, but we could eventually in that issue. VR180 support will require a lot more investigation and likely coordination w/ other browser vendors.

**Comment by cvan** [comment]
@MortimerGoro ready for another review, if you have a min

**Comment by bluemarvin** [comment]
I just tested this patch with https://www.youtube.com/watch?v=hFjzFSidX3s&mozVideoProjection=360_auto and it default to 360p resolution.

**Comment by MortimerGoro** [comment]
I see this error in the console

`[JavaScript Error: "SyntaxError: missing = in const declaration" {file: "moz-extension://a9fe3216-9463-45c6-bf52-4eb51c350ea4/main.js" line: 69 column: 10 source: "    const prefs.qualities = ["}]`

**Comment by MortimerGoro** [comment]
If fixed the SyntaxError but it still doesn't work.

@cvan I see that you have changed the code to user player.setPlaybackQuality(). I already tried that the first time we added the webextension and added some comments there. I found that `setPlaybackQuality` is  a no-op in the current Youtube API. It's mentioned in the docs:

```
setPlaybackQuality is now considered a "no-op"; calling this function will not change the player behavior. The player will use a variety of signals to determine the optimal playback quality.
Users are able to manually request a specific playback quality via the quality selector in the player controls.
```

So I think the only way would be to try the fake touches again

**Comment by cvan** [comment]
@MortimerGoro:

> @cvan I see that you have changed the code to use player.setPlaybackQuality(). I already tried that the first time we added the webextension and added some comments there. I found that `setPlaybackQuality` is  a no-op in the current Youtube API. It's mentioned in the docs:

Yeah, I can actually remove the [`player.setPlaybackQuality(newBestQuality)` line](https://github.com/MozillaReality/FirefoxReality/pull/1052/files#diff-44d6081a68b4885b80582ec64fc90a4dR144).

But what I discovered is that you can use [`player.setPlaybackQualityRange(newBestQuality, newBestQuality)`](https://github.com/MozillaReality/FirefoxReality/pull/1052/files#diff-44d6081a68b4885b80582ec64fc90a4dR150), and that actually seems to work.

I've been able to run my branch successfully on YouTube videos on desktop Firefox + Chrome, and the default minimum quality is changed or if specified by the `vq=` or `quality=` query-string parameter, that quality gets enforced.

Where I'm stuck on:

- What's made this branch difficult is the loading of the content script via GV's Web Extension API seems to load differently than desktop Firefox's. I might want to change `document_start` to `document_end` in the `manifest.json`.
- The firing of the `onStateChange` event from YouTube's video player isn't fired on FxR but is on Firefox desktop w/ the same UA (w/ `disable_polymer=1` correctly set in the URL) and Chrome too.
- For some reason (likely user-activation requirement or YouTube's policies for progressively limiting the experience based on screen size/bandwidth/performance/etc.), YouTube doesn't begin playing videos, and _most_ times [`player.getAvailableQualityLevels()` isn't available](https://github.com/MozillaReality/FirefoxReality/pull/1052/files#diff-44d6081a68b4885b80582ec64fc90a4dR60). I use `player.getAvailableQualityLevels()` to cross-reference with all the supported YouTube quality levels to identify which index in the array to select to change the quality. There's also code to find [the next best quality level](https://github.com/MozillaReality/FirefoxReality/pull/1052/files#diff-44d6081a68b4885b80582ec64fc90a4dR120) if `1440p` is not available (or the forced value in the query-string parameter of the URL).

To overcome these issues, I think I'm going to have to stop relying on these proprietary YouTube events to know when we are ready to set the quality.

I'll likely have to use a `setInterval` loop every couple milliseconds to keep checking until the player is ready and the `player.getAvailableQualityLevels()` actually returns the video's quality (which seems to be only after the video starts playing, which requires a user-activation click to play the video).

If anyone has any better approaches here, let me know.

As you can imagine, the development process of making changes with Android Studio -> Oculus Go every time there's a JS change is slow and difficult to remotely debug via desktop Firefox's `about:debugging` and the WebIDE.

Earlier, I did some experimenting with injecting a script hosted elsewhere so we'd be able to update the content script for YouTube whenever without having to ship a new release with the hardcoded content script.

@bluemarvin expressed concerns that because GeckoView's WebExtension API is not intended for public/end-developer consumption yet, it's likely unreasonable or irresponsible to inject a third-party script (even if it's ours and hosted on a Mozilla-owned domain). (CSP policies can be enforced, but still that's a possible MITM attack vector.)

The precedent I've seen in Gecko, specifically for Android WebCompat issues, is a hardcoded [extension that's stored in `mozilla-central`](https://searchfox.org/mozilla-central/source/mobile/android/extensions/webcompat) and bootstrapped into Firefox. Every new adjustment requires a new Firefox release, but it seems to not been an issue from the bugs and commit activity I read. @miketaylr likely has some recommendations of best practices here for a scalable, realistic solution for maintaining WebCompat content scripts in FxR.

To speed up development, I've been writing the code and running it from `about:addons` on desktop Firefox, and that's helped me iterate faster.

Writing the hacks has been fun, and I still have the old code for manually pressing the buttons in the DOM. So we could go back to that if the [`player.setPlaybackQualityRange(newBestQuality, newBestQuality)`](https://github.com/MozillaReality/FirefoxReality/pull/1052/files#diff-44d6081a68b4885b80582ec64fc90a4dR150) code doesn't run properly in FxR. (I got it all working in FxR several times but accidentally broke it. Fragile stuff, but I'll battle-test this and make sure it always works.)

This patch is in a broken state right now because I've rewritten it so many times. In the meantime, in between coordinating the v1.1.4 release and other work tasks, I'm going to prioritise working on this to get it consistently stable and ready to review + land. If anyone has recommendations on the above items, I'd appreciate the help and will adjust the patch accordingly. Thanks, everyone.

**Comment by MortimerGoro** [comment]
I think `setInterval` is the only realiable way right now to wait for quality availabilities. I'm ok with using that for now

**Comment by MortimerGoro** [comment]
just tried `player.setPlaybackQualityRange(newBestQuality, newBestQuality)` using remote debugger in FxR and desktop and it worked in both. Good catch! So if we make the load detection more reliable with setInterval it should be ready to land

**Comment by bluemarvin** [comment]
I've updated the PR to use the methods discussed. Please give it a try.

**Comment by MortimerGoro** [comment]
It works. I'm going to merge this for the release

**Comment by cvan** [comment]
thanks for the assists! much appreciated :heart:

File: app/src/main/assets/web_extensions/youtube_webcompat/main.css [hunk 2]
```diff
@@ -1,3 +1,6 @@
-div.alert-with-button-renderer {
-	display:none;
+/* To hide harmless warnings. */
+.alert-with-button-renderer, .ytm-alert-with-button-renderer, ytm-alert-with-button-renderer,
+.ytp-ad-module, ytp-ad-module
```
Comment (cvan) [inline app/src/main/assets/web_extensions/youtube_webcompat/main.css]: oops, missing comma (regressed issue #837 and #1056)

Comment (cvan) [reply app/src/main/assets/web_extensions/youtube_webcompat/main.css]: fixed in PR #1217: https://github.com/MozillaReality/FirefoxReality/pull/1217/files#diff-ebdf9c97fadbb79c463a9b4302a68491R3

**Pull request merged**